attempt to index a nil value (field 'tabs') -- quarto webite

### Bug description

When building a Quarto website that includes blog posts (https://geocompx.org/), rendering fails with recent Quarto versions (1.7 and 1.8) due to a Lua filter error on one of the blog posts. The error occurs even though the post is frozen and should not be re-evaluated during rendering. This behavior breaks the entire site build process.

### Steps to reproduce

The complete source code of the website together with the related GitHub action can be found at https://github.com/geocompx/geocompx.org. I also created much simplified version of the website with just this one blog post, which you can find at https://github.com/Nowosad/quarto-tmp.



### Actual behavior

```
Error running filter /opt/quarto/1.8.4/share/filters/main.lua:
/opt/quarto/1.8.4/share/filters/main.lua:21768: attempt to index a nil value (field 'tabs')
stack traceback:
        /opt/quarto/1.8.4/share/filters/main.lua:935: in function </opt/quarto/1.8.4/share/filters/main.lua:934>
        (...tail calls...)
        /opt/quarto/1.8.4/share/filters/main.lua:1152: in function </opt/quarto/1.8.4/share/filters/main.lua:1147>
        (...tail calls...)
        /opt/quarto/1.8.4/share/filters/main.lua:2943: in function </opt/quarto/1.8.4/share/filters/main.lua:2937>
        (...tail calls...)
        ...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:34: in upvalue 'run_filter_function'
        ...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:223: in function <...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:202>
        (...tail calls...)
        [C]: in function 'pandoc.List.map'
        ...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:125: in upvalue 'recurse'
        ...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:222: in local 'jogger'
        ...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:171: in upvalue 'recurse'
        ...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:222: in function <...1.8.4/share/pandoc/datadir/../../filters/modules/jog.lua:202>
        (...tail calls...)
        /opt/quarto/1.8.4/share/filters/main.lua:572: in function </opt/quarto/1.8.4/share/filters/main.lua:550>
        (...tail calls...)
        /opt/quarto/1.8.4/share/filters/main.lua:1318: in local 'callback'
        /opt/quarto/1.8.4/share/filters/main.lua:1336: in upvalue 'run_emulated_filter_chain'
        /opt/quarto/1.8.4/share/filters/main.lua:1372: in function </opt/quarto/1.8.4/share/filters/main.lua:1369>
WARN: Should not have arrived here:
ERROR: [non-error-thrown] undefined

Stack trace:
    at _Command.handleError (file:///opt/quarto/1.8.4/bin/quarto.js:10023:117)
    at _Command.parseCommand (file:///opt/quarto/1.8.4/bin/quarto.js:9710:12)
    at eventLoopTick (ext:core/01_core.js:178:7)
    at async quarto (file:///opt/quarto/1.8.4/bin/quarto.js:164259:5)
    at async file:///opt/quarto/1.8.4/bin/quarto.js:164287:5
    at async file:///opt/quarto/1.8.4/bin/quarto.js:164142:14
    at async mainRunner (file:///opt/quarto/1.8.4/bin/quarto.js:164144:5)
    at async file:///opt/quarto/1.8.4/bin/quarto.js:164280:3
```

### Expected behavior

Rendered website.

### Your environment

- OS: Fedora 42

### Quarto check output

```bash
Quarto 1.8.4
[✓] Checking environment information...
      Quarto cache location: /home/user/.cache/quarto
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.6.3: OK
      Dart Sass version 1.87.0: OK
      Deno version 2.3.1: OK
      Typst version 0.13.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.8.4
      Path: /opt/quarto/1.8.4/bin

[✓] Checking tools....................OK
      TinyTeX: v2024.07.03
      Chromium: 869685

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /home/user/.TinyTeX/bin/x86_64-linux
      Version: 2024

[✓] Checking Chrome Headless....................OK
      Using: Chrome found on system
      Path: /usr/bin/google-chrome
      Source: PATH

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.13.3
      Path: /usr/bin/python3
      Jupyter: 5.7.2
      Kernels: python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.5.0
      Path: /usr/lib64/R
      LibPaths:
        - /home/user/R/x86_64-redhat-linux-gnu-library/4.5
        - /usr/lib64/R/library
        - /usr/share/R/library
      knitr: 1.50
      rmarkdown: 2.29

[✓] Checking Knitr engine render......OK
```

Yes -- it is in GitHub actions (https://github.com/geocompx/geocompx.org/actions/runs/15002379066) -- version 1.7.31:

```bash
22/46] post/2023/ogh23/index.qmd
Error running filter /opt/quarto/share/filters/main.lua:
/opt/quarto/share/filters/main.lua:21750: attempt to index a nil value (field 'tabs')
stack traceback:
	/opt/quarto/share/filters/main.lua:935: in function </opt/quarto/share/filters/main.lua:934>
	(...tail calls...)
	/opt/quarto/share/filters/main.lua:1152: in function </opt/quarto/share/filters/main.lua:1147>
	(...tail calls...)
	/opt/quarto/share/filters/main.lua:2943: in function </opt/quarto/share/filters/main.lua:2937>
	(...tail calls...)
	...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:34: in upvalue 'run_filter_function'
	...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:223: in function <...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:202>
	(...tail calls...)
	[C]: in function 'pandoc.List.map'
	...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:125: in upvalue 'recurse'
	...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:222: in local 'jogger'
	...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:171: in upvalue 'recurse'
	...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:222: in function <...uarto/share/pandoc/datadir/../../filters/modules/jog.lua:202>
	(...tail calls...)
	/opt/quarto/share/filters/main.lua:572: in function </opt/quarto/share/filters/main.lua:550>
	(...tail calls...)
	/opt/quarto/share/filters/main.lua:1318: in local 'callback'
	/opt/quarto/share/filters/main.lua:1336: in upvalue 'run_emulated_filter_chain'
	/opt/quarto/share/filters/main.lua:1372: in function </opt/quarto/share/filters/main.lua:1369>
ERROR: Error
    at renderFiles (file:///opt/quarto/bin/quarto.js:86948:29)
    at eventLoopTick (ext:core/01_core.js:175:7)
    at async renderProject (file:///opt/quarto/bin/quarto.js:87354:25)
    at async renderForPublish (file:///opt/quarto/bin/quarto.js:118724:33)
    at async renderForPublish (file:///opt/quarto/bin/quarto.js:114045:24)
    at async Object.publish1 [as publish] (file:///opt/quarto/bin/quarto.js:114530:26)
    at async publishSite (file:///opt/quarto/bin/quarto.js:118762:38)
    at async publish7 (file:///opt/quarto/bin/quarto.js:118983:61)
    at async doPublish (file:///opt/quarto/bin/quarto.js:118943:13)
    at async publishAction (file:///opt/quarto/bin/quarto.js:118954:9)
Error: Process completed with exit code 1.
```

--- a/news/changelog-1.8.md
+++ b/news/changelog-1.8.md
@@ -35,3 +35,7 @@ All changes included in 1.8:
 ## Crossrefs
 
 - ([#12615](https://github.com/quarto-dev/quarto-cli/pull/12615)): Adds `algorithm` to theorem environments. (author: @jeremy9959)
+
+## Lua Filters
+
+- ([#12727](https://github.com/quarto-dev/quarto-cli/issues/12727)): Do not crash in the presence of malformed tabset contents.

--- a/src/resources/filters/customnodes/panel-tabset.lua
+++ b/src/resources/filters/customnodes/panel-tabset.lua
@@ -146,6 +146,10 @@ _quarto.ast.add_handler({
 
   constructor = function(params)
     local node = _quarto.ast.create_custom_node_scaffold("Tabset", "Block")
+    if params.tabs == nil then
+      warn("No tabs found in tabset. Please check that your markdown includes tab headings as appropriate.")
+      params.tabs = pandoc.List()
+    end
 
     local custom_data = {
       __quarto_custom_node = node,

--- a/tests/docs/smoke-all/2025/05/13/issue-12727.qmd
+++ b/tests/docs/smoke-all/2025/05/13/issue-12727.qmd
@@ -1,0 +1,35 @@
+---
+title: "Geographic data analysis in R and Python: comparing code and outputs for vector data"
+---
+
+::: {.panel-tabset group="language"}
+
+## Python
+
+```python
+print("ok")
+```
+
+## R
+
+```r
+print("hello")
+```
+
+:::
+
+::: {.panel-tabset group="language"}
+
+## Python
+
+```python
+print("Missing the headers here but we shouldn't crash")
+```
+
+## R
+
+```r
+cat("hello")
+```
+
+:::

